Flatpak

jwillikers opened the pull request:
This PR makes the necessary changes to support Flatpak.

It adds an initial AppStream metadata file for the project will be used to display the project in Software Stores.
Currently, I've worked around the screenshot being a bit too large by putting a resized one in my PR.
It appears that newer versions of AppStream support larger resolutions for screenshots, so maintaining this separate screenshot should only be required temporarily.

Due to issues with SerialPort in AVRGirl, I have introduced a flashing mechanism which uses `avrdude` instead when in the Flatpak.
The reason for the trouble is that SerialPort relies on Udev to detect serial ports.
Udev is not available in Flatpak's, so I instead patch out direct uses of SerialPort with queries to `sysfs` instead to detect the serial ports within the Flatpak.
Udev permissions should be configured outside of the Flatpak appropriately so that the user has permission to access the serial ports of any Keyboards.
This usually involves adding the user to the necessary group such as `dialout` or `uucp`.

To get Chrysalis on Flathub, a PR will need to be opened against [flathub/flathub](https://github.com/flathub/flathub) with a release version of the Flatpak manifest.
Someone from the Chrysalis project should submit the PR to Flathub to ensure they have write access to the repository when it is created.
I have a branch ready to be submitted to Flathub [here](https://github.com/jwillikers/flathub/tree/io.keyboard.Chrysalis).
Once this PR is merged, I will update that branch to track Chrysalis proper instead of this branch.
After which, my branch should be forked by a Chrysalis member and submitted to [flathub/flathub](https://github.com/flathub/flathub) targeting the `new-pr` branch.

I've outlined the remaining work below.

Tasks
- [X] Fix device detection through serialport when inside the Flatpak, which does not have `udevadm` available. Ideally, fix this issue upstream. See issue serialport/node-serialport#2306. I've Patched this in Chrysalis since it wasn't too difficult.
- [X] Determine if Chrysalis uses the native file chooser dialogs to import / export layers. If it doesn't, then access to certain user directories must be allowed in the Flatpak's permissions. Zypak doesn't currently support those, so filesystem access is restricted to the user's Downloads directories by default.
- [X] Disable installing the Udev rules when running inside the Flatpak.
- [X] Cleanup any lingering files in the Flatpak which are not necessary.
- [X] Verify the Flatpak works on `aarch64`.
- [X] Look into enabling Zypak to handle the chrome-sandbox, but *without* blocking device access in `/dev`. I've managed to at-least use the Zypak wrapper script, though I set an environment variable to disable Chrome's sandbox.
- [x] Ensure the release process verifies the AppStream metadata and `.desktop` files to avoid issues on Flathub.
- [x] Fix infinite, visual loop during step 3 of firmware upgrades for Arduino-based boards, which uses a library that lists devices with Serialport, which doesn't work in the Flatpak.
- [ ] Submit a Pull Request to [flathub/flathub](https://github.com/flathub/flathub) with a *Release* version of the Flatpak manifest to get published.

jwillikers commented:
@algernon I've got the Flatpak building, but I'm getting the dreaded white-screen of death when I run it. It's probably an initialization issue of some sort. Any ideas on the easiest way to get more logging output or otherwise debug the issue?

algernon commented:
> @algernon I've got the Flatpak building, but I'm getting the dreaded white-screen of death when I run it. It's probably an initialization issue of some sort. Any ideas on the easiest way to get more logging output or otherwise debug the issue?

Can you open the developer console (ctrl+shift+i), and see if there's anything there?

jwillikers commented:
Yup, that told me exactly what was wrong. It's getting there. I've got the white screen fixed and will be moving on to fix the missing window decorations in GNOME and the fact that it can't actually detect my Model 01 yet.

jwillikers commented:
The missing title bar is only an issue running under Wayland, so I've configured the Flatpak to use X11 but to use Wayland if the user enables the Wayland socket for the Flatpak.

Now I'll just have to figure out why Chrysalis can't detect the keyboard from inside the Flatpak.

jwillikers commented:
Disabling the sandbox allows access to the `tty` devices under `/dev` in the Flatpak, but it looks like the serialport library uses the command `udevadm` to determine what serialport devices are available on Linux. That's why it can't find my Model 01, there is no `udevadm` binary in the Flatpak.

This is probably going to be pretty painful to workaround, though I've managed to install `eudev` inside the Flatpak to provide `udevadm`, it's not giving me the same output for `/dev/ttyACM0` inside the Flatpak as it is outside.

jwillikers commented:
Well, if I can't the easiest way to do this will be to just query the filesystem directly in the Flatpak.
The `tty` devices under `/sys/class` contain the information required under the first two fields for the `PRODUCT` key.
The example below is from inside the Chrysalis Flatpak.

```
$ cat /sys/class/tty/ttyACM0/device/uevent
DEVTYPE=usb_interface
DRIVER=cdc_acm
PRODUCT=1209/2301/100
TYPE=239/2/1
INTERFACE=2/2/0
```

But running `udevadm` from the eudev package doesn't give me the product model / manufacturer:

```
$ udevadm info -q all /dev/ttyACM0
P: /devices/pci0000:00/0000:00:14.0/usb1/1-5/1-5.3/1-5.3.3/1-5.3.3:1.0/tty/ttyACM0
N: ttyACM0
E: DEVNAME=/dev/ttyACM0
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-5/1-5.3/1-5.3.3/1-5.3.3:1.0/tty/ttyACM0
E: MAJOR=166
E: MINOR=0
E: SUBSYSTEM=tty
```

Perhaps `eudev` just isn't installed correctly.

jwillikers commented:
I've opened issue serialport/node-serialport#2306 in the hope of maybe getting this fixed upstream.

algernon commented:
At some point in the not too distant future, we're likely going to switch away from node-serialport, and use WebSerial instead, which should fix the detection problem too, among other things. But even if that's in the not too distant future, it's still gonna be a while.

jwillikers commented:
Thanks, for the heads up on that, and that sounds like a good long-term solution. I wasn't aware of WebSerial, either, and that's really cool. It also sounds like a bit of an undertaking to change to that! Hopefully I'll be able to get a simple workaround implemented before too long.

jwillikers commented:
@algernon Got it! It took me a while since I'm pretty clueless when it comes to Javascript, but the algorithm itself is pretty straightforward. Luckily, we only list TTY devices in two places, so I just patched the code to use the workaround I described previously when it detects it is inside a Flatpak. It just iterates over the files in `/sys/class/tty` and extracts the necessary data.

The patch works great and luckily we don't need to patch anything else. The only grievance is that Electron doesn't use the native filesystem portal, so we have to grant filesystem access piecewise so that users can import and export layout files. For now, I've just enabled access to the user's `~/Downloads` directory by default, which I think is a pretty safe allowance. Hopefully Zypak will reintroduce support for enabling Electron to use the native file chooser in the future.

I also disabled the option to install Udev rules when inside the Flatpak since this won't make any difference. Hopefully, I did this correctly.

I should be able to wrap-up any outstanding items from my checklist tomorrow. I'll test building the Chrysalis Flatpak on my `aarch64` Pinebook Pro. Otherwise, I think the primary thing remaining is that we make sure CI is configured to your liking for verifying the AppStream metadata and `.desktop` files. Once that (and everything else that comes up ;) is settled, I'll create a template repository for you to submit in a PR to Flathub.

jwillikers commented:
I've got some good news and some bad news.

The bad news is that because the library used to connect to the Arduino devices uses Serialport to list devices which leads to an infinite spinning loop (and error) during the flashing portion of firmware upgrades, though the actual update appears to complete successfully. There's an exception thrown, which should probably be caught and handled anyways.

The good news is that everything just works on `aarch64` and there were no issues getting that to work!

jwillikers commented:
@algernon I've updated the Flatpak branch to use the latest Flatpak runtime and Nodejs LTS. Firmware updates through AVRGirl are still not possible due to the lack of Udev in the Flatpak, though layout modifications are still possible in Chrysalis and upload to the Keyboard without a problem. Everything seems to be working really well. Heck, it appears that I've even got it running under Wayland now. That's really cool to see.

I think firmware upgrades will be possible for the `Model100` since it uses `dfu-util` and not `AVRGirl`. I'm still waiting for my Model100 to test, though. I removed the pre-compiled binaries from the Flatpak and build `dfu-util` from source as part of the Flatpak. This reduces the size bloat from the binaries for other platforms and allows me to build for `aarch64` for which no pre-built binary is included.

razzeee commented on src/renderer/utils/flatpak.js:
```diff
@@ -0,0 +1,119 @@
+async function insideFlatpak() {
+  return (
+    process.platform == "linux" &&
+    (await fs.promises.access(flatpakInfoPath).catch(() => {
```
This doesn't seem to work for me, ended up with 
```
async function insideFlatpak() {
  return process.platform === 'linux' && fs.existsSync(flatpakInfoPath);
}
```

jwillikers replied:
I have no clue what I'm doing when it comes to Javascript, so not too surprised it didn't work. I've update this as you've suggested. Thanks!

razzeee commented on src/renderer/utils/flatpak.js:
```diff
@@ -0,0 +1,119 @@
+      let fileStream;
+      try {
+        fileStream = await createReadStreamSafe(
+          path.join(dirPath, "device", "uevent")
```
I don't seem to have a `device` folder in between?

razzeee replied:
Oh, it only exists for some of the devices

jwillikers replied:
Yes, there's a bunch of serial device directories mucking everything up in there, but the ones with an actual `device` folder are _real_.

razzeee replied:
Yeah, but you get rid of most of them, which is not the behavior I'm seeing with what I'm trying to port.

So I changed it to keep those, if the `device` folder exists and only add the vendor/product when I have them.

jwillikers commented:
@razzeee Thanks for taking the time the to review this PR, I really appreciate it. Let me know if you run into anything else.

jwillikers commented:
> I'll try to have a look at it sometime soon. Great work @jwillikers, thank you!
> 
> If it is possible to use the external `dfu-util` tool from flatpak, perhaps we can do the same for AVR boards, and use `avrdude` there, as a workaround. But I'll have a closer look once I get around to actually testing this.

I had the exact same thought, but didn't want to necessarily use `avrdude` only for the Flatpak in case that would diverge too much. Since `avrdude` is a reasonably well-written CMake project and looks to only require `libusb`, building from source shouldn't be a problem. Would just need to create the flashing interface like there is for `dfu-util`. Just let me know what you end up thinking. Cheers!

razzeee commented:
Here's my version written in typescript, please let me know, if you think this is  licensing problem, as I've based it heavily on your code.

https://github.com/razzeee/ExpressLRS-Configurator/blob/device-list-flatpak/src/api/src/flatpak.ts

jwillikers commented:
@algernon I've got the Flatpak using `avrdude` for flashing. I've successfully test it with both the Model01 and Atreus2.

jwillikers commented:
Thanks for the update, I understand!

jwillikers commented:
I've got a Model100 now, but firmware upgrades are a real problem for it. It gets hung up after switching the Model100 into bootloader mode. I added some logging of what USB devices are detected and even though the Model100 is in bootloader mode, the old id when not in bootloader mode is reported, i.e. `6`. Checking `/sys/bus/usb/devices/*/idProduct`, I can see the correct ID, `0005`, from inside the Flatpak when the Model100 is flipped into bootloader mode. I think there's something wrong with USB device detection in the Flatpak. Perhaps the lack of `udev` in the Flatpak is a problem for `node-usb`?

jwillikers commented:
Ha! Getting upgrades to work ended up being surprisingly similar to how I modified Chrysalis to detect serial ports in the Flatpak. Firmware upgrades for the Model100 are now fully functional inside the Flatpak.

obra commented:
Chrysalis now runs in-browser using WebSerial and WebUSB, so I'm closing out this PR as obsolete.
